Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81763
* #81762
* __->__ #81761

`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

cc @ezyang @gchanan @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @ngimel @fdrocha

lezcano committed:
Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm

`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

Lezcano commented:
This one would need a matching PR on the functorch side. We should ping Richard once it's approved.

Lezcano commented:
We would also need to open an ONNX issue.

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

ngimel commented on torch/functional.py:
```diff
@@ -1439,6 +1440,42 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
+        if isinstance(p, str):
+            if p == "fro" and (dim is None or isinstance(dim, int) or len(dim) <= 2):
+                if out is None:
+                    return torch.linalg.vector_norm(input, 2, _dim, keepdim, dtype=dtype)
```
I guess linalg.vector_norm and matrix_norm refs would need to be registered as decompositions to support `torch.norm` in functorch? cc @chillee, @zou3519 Are there functorch tests that would make sure that this doesn't break things?

Chillee replied:
No i don't think this would be an issue with vmap. We already have a batching rule for `vector_norm`.

Lezcano replied:
According to [this failing test](https://github.com/pytorch/pytorch/runs/7469343187?check_suite_focus=true), we'll need to touch a bit some part of functorch, probably adding a batching rule for `matrix_norm`. Now that functorch is merged in core, I can do that in this PR directly.

zou3519 replied:
Yeah the functorch tests in PyTorch CI should give signal to this

Lezcano replied:
I added the decomposition for `matrix_norm`, but now I have a new error popping up:
https://github.com/pytorch/pytorch/runs/7862792767?check_suite_focus=true
I would xfail it, but I think it is relevant, as it may indicate that the decomposition is not correct? WDYT @zou3519?

zou3519 replied:
cc @samdow -- can we xfail this? matrix_norm requires the input to have two or more dimensions so I think this test doesn't apply

samdow replied:
So that test is checking that with a 1D input to the linalg functions, we get an error. So (1) it's good that we're still getting an error (🥳 ) and (2) it's weird that the error doesn't mention the `assertRaisesRegex` context manager that it's under

If you xfail it (and leave a comment that just says like "erroring but still fails test" please!), I'll look at it more since that's irrelevant to you

Lezcano replied:
It's not caught because the error type that's thrown `IndexError` does not match the error being caught. Sure, I'll xfail it for now. I think the fix should be quite simple for a person that understands functorch.

Lezcano replied:
FWIW, this xfail is not necessary anymore. Whatever was causing it has been fixed.

Lezcano replied:
Actually, it passes locally and in some CI jobs, and it fails in others. I'm skipping it.

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

Lezcano commented:
@ngimel this one should be ready for review.
I had to skip some ONNX tests, and I filled up an issue to note this: https://github.com/pytorch/pytorch/issues/83661

ngimel commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -411,6 +411,9 @@ TORCH_LIBRARY_IMPL(aten, Autocast, m) {
                                  std::tuple<Tensor,Tensor,Tensor> (const Tensor&, IntArrayRef, const c10::optional<Tensor>&, const c10::optional<Tensor>&, double),
                                  &ADD_NS(native_layer_norm)>::type::call)));
   KERNEL(ADD_NS(group_norm), "group_norm", Tensor (const Tensor &, int64_t, const c10::optional<Tensor>&, const c10::optional<Tensor>&, double, bool), fp32)
+  KERNEL(ADD_NS(linalg_vector_norm), "linalg_vector_norm", Tensor (const Tensor &, const Scalar&, at::OptionalIntArrayRef, bool, c10::optional<c10::ScalarType>), fp32_set_opt_dtype)
```
I think this change is not needed, we shouldn't produce fp32 output in autocast unless user specifically requests it

Lezcano replied:
I must have added this because some test was failing. Let me remove it see what happens.

ngimel replied:
If the tests weren't failing before, just redispatch shouldn't have changed it. Adding functions to convert to fp32 list is expensive.

Lezcano replied:
This is the failing test: https://github.com/pytorch/pytorch/runs/8049988791?check_suite_focus=true

ngimel replied:
How was it working previously? What caused it to produce fp32 output?

Lezcano replied:
Most probably because of these lines:
https://github.com/pytorch/pytorch/blob/ba8080a0cf0511365aef735f30c6e209baceead5/aten/src/ATen/autocast_mode.cpp#L463-L468
and similar lines for `frobenius_norm` and `nuclear_norm`.

ngimel replied:
Ok, the `fp32_set_opt_dtype` makes sense

zou3519 commented on functorch/test/test_vmap.py:
```diff
@@ -3759,6 +3759,9 @@ def f(e_):
         skip('linalg.multi_dot'),  # accepts list of tensor inputs, has its own special test
         xfail('linalg.vander'),
         xfail('linalg.vecdot'),
+        # Throws in vmap
+        # IndexError: Dimension out of range (expected to be in range of [-1, 0], but got -2)
+        xfail('linalg.matrix_norm'),
```
This PR didn't add `linalg.matrix_norm`, right? If so then this is a regression that we should fix before merging

ngimel replied:
This test is xfailed per @samdow's advice https://github.com/pytorch/pytorch/pull/81761#discussion_r948021867

zou3519 replied:
Oh, I forgot about the discussion above. This is fine, thanks for checking @ngimel.

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

BowenBao commented:
@Lezcano you can fix these tests by running `python test/onnx/test_operators.py --accept`, and add the updated expect file into this commit. 
```
         - test/onnx/test_operators.py:955 TestOperators.test_frobenius_norm
         - test/onnx/test_operators.py:652 TestOperators.test_norm_p1
         - test/onnx/test_operators.py:656 TestOperators.test_norm_p2
```
But feel free to skip the tests if you are unsure of the changes.

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

Lezcano commented on test/onnx/test_utility_funs.py:
```diff
@@ -186,7 +186,6 @@ def forward(self, x):
 
         for node in graph.nodes():
             self.assertNotEqual(node.kind(), "onnx::ReduceL2")
-        self.assertEqual(len(list(graph.nodes())), 2)
```
@BowenBao thank you, I applied your suggestions to the failing expecttests. I also needed to remove this assert equal give that, since now `norm` is composite, it so happens that the length of the decomposition changes between ONNX versions (it changes from version 12 to 13). I hope that this is fine.

Lezcano commented:
Running `python test/onnx/test_operators.py --accept` was not good enough, as it's not failing in CI. I'm going to revert the changes and keep the skips.

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

BowenBao commented:
@Lezcano I see more tests are failing in `test/onnx/test_pytorch_onnx_onnxruntime.py`. Please update the linked issue as you skip those as well, we'll take a look.

lezcano committed:
Update on "Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm"


`torch.norm` is very odd. Some notable issues are:

- The default value of `"fro"` in `torch.norm` has an odd behaviour when `dim=None`. This is handled in the new dispatch
- The treatment of the `dtype` argument in `torch.norm` was completely wrong. This should fix it
- Some `out=` variants in the previous implementation were also wrong. This should fix those.
- This new dispatch should make some paths much faster. For example, `torch.norm(x)` where `x` is complex.

I'll try to make the changes in these PRs as incremental as possible as this is a tricky one.

[ghstack-poisoned]

Lezcano commented:
@ngimel I think the CI should be (at last) green, PTAL.

Lezcano commented:
@pytorchbot merge

mehtanirav commented:
Reverting this PR as it is causing test failures in pytorch/glow


```
FAIL: test_norm_3d_inner_axis (glow.torch_glow.nodes.norm_test.TestNorm)
Basic test of the PyTorch norm Node on Glow.

AssertionError: Expected fusion of {'aten::frobenius_norm'}, but only set() was fused in graph
graph(%self : __torch__.glow.torch_glow.nodes.norm_test.SimpleNormModule,
      %tensor : Float(*, *, *, requires_grad=0, device=cpu)):
  %8 : int[] = prim::Constant[value=[1]]()
  %2 : NoneType = prim::Constant()
  %3 : bool = prim::Constant[value=0]() # /data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/aab7ed39/glow/glow/torch_glow/tests/norm_test#binary,link-tree/torch/functional.py:1471:0
  %5 : int = prim::Constant[value=2]() # /data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/aab7ed39/glow/glow/torch_glow/tests/norm_test#binary,link-tree/torch/functional.py:1471:0
  %7 : Tensor = aten::linalg_vector_norm(%tensor, %5, %8, %3, %2) # /data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/aab7ed39/glow/glow/torch_glow/tests/norm_test#binary,link-tree/torch/functional.py:1471:0
  return (%7)
```

These tests are defined in https://github.com/pytorch/glow/blob/1b4e9d1fdcb0fc8abf7a6be26a781256bc307b17/torch_glow/tests/nodes/norm_test.py#L50

mehtanirav commented:
@pytorchbot revert -m "Breakages in pytorch/glow" -c "ghfirst"

ngimel commented:
Do you know who should we talk to on the glow side?

Lezcano commented:
@kit1980 trying to reland this PR here:
https://github.com/pytorch/pytorch/pull/84624
It'll be best to import it to fabricator before relanding it for obvious reasons.
